frame/bags-list/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+		Rebagged(T::AccountId, VoteWeight, VoteWeight),
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Declare that some `dislocated` account has, through rewards or penalties, sufficiently
+		/// changed its weight that it should properly fall into a different bag than its current
+		/// one.
+		///
+		/// Anyone can call this function about any potentially dislocated account.
+		///
+		/// Will never return an error; if `dislocated` does not exist or doesn't need a rebag, then
+		/// it is a noop and fees are still collected from `origin`.
+		#[pallet::weight(T::WeightInfo::rebag())]
+		pub fn rebag(origin: OriginFor<T>, dislocated: T::AccountId) -> DispatchResult {
+			ensure_signed(origin)?;
+			let current_weight = T::VoteWeightProvider::vote_weight(&dislocated);
+			let _ = Pallet::<T>::do_rebag(&dislocated, current_weight);
+			Ok(())
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn integrity_test() {
+			// ensure they are strictly increasing, this also implies that duplicates are detected.
+			assert!(
+				T::BagThresholds::get().windows(2).all(|window| window[1] > window[0]),
+				"thresholds must strictly increase, and have no duplicates",
+			);
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Move an account from one bag to another, depositing an event on success.
+	///
+	/// If the account changed bags, returns `Some((from, to))`.
+	pub fn do_rebag(
+		account: &T::AccountId,
+		new_weight: VoteWeight,
+	) -> Option<(VoteWeight, VoteWeight)> {
+		// if no voter at that node, don't do anything.
+		// the caller just wasted the fee to call this.
+		let maybe_movement = list::Node::<T>::get(&account)
+			.and_then(|node| List::update_position_for(node, new_weight));
+		if let Some((from, to)) = maybe_movement {
+			Self::deposit_event(Event::<T>::Rebagged(account.clone(), from, to));
+		};
+		maybe_movement
+	}
+
+	/// Equivalent to `ListBags::get`, but public. Useful for tests in outside of this crate.
+	#[cfg(feature = "std")]
+	pub fn list_bags_get(weight: VoteWeight) -> Option<list::Bag<T>> {
+		ListBags::get(weight)
+	}
+}
+
+impl<T: Config> SortedListProvider<T::AccountId> for Pallet<T> {
+	type Error = Error;
+
+	fn iter() -> Box<dyn Iterator<Item = T::AccountId>> {
+		Box::new(List::<T>::iter().map(|n| n.id().clone()))
+	}
+
+	fn count() -> u32 {
+		CounterForListNodes::<T>::get()
+	}
+
+	fn contains(id: &T::AccountId) -> bool {
+		List::<T>::contains(id)
+	}
+
+	fn on_insert(id: T::AccountId, weight: VoteWeight) -> Result<(), Error> {
+		List::<T>::insert(id, weight)
+	}
+
+	fn on_update(id: &T::AccountId, new_weight: VoteWeight) {
+		Pallet::<T>::do_rebag(id, new_weight);
+	}
+
+	fn on_remove(id: &T::AccountId) {
+		List::<T>::remove(id)
+	}
+
+	fn regenerate(
+		all: impl IntoIterator<Item = T::AccountId>,
+		weight_of: Box<dyn Fn(&T::AccountId) -> VoteWeight>,
+	) -> u32 {
+		List::<T>::regenerate(all, weight_of)
+	}
+
+	fn sanity_check() -> Result<(), &'static str> {
```

Comment:
should be feature gated as well.

Reply:
3594eab2231198209b0c45b66e2c761707655260